### PR TITLE
add validate function. add exceptions in parallel_process

### PR DESCRIPTION
**Pull request by nitsandaniel1**

### CODE REVIEW

File: processing.py [hunk 1]
```diff
@@ -234,21 +234,29 @@ def convert_fastq_to_fasta(output_dir, fastq_file):
...
+        basecall_output = os.path.join(output_dir, 'basecall')
+        if not quality_threshold:
+            quality_threshold = 30
+        os.makedirs(basecall_output, exist_ok=True)
+        basecall(blast_output_file=blast_output_file, fastq_file=fastq_file, output_dir=basecall_output,
              quality_threshold=quality_threshold, mode=mode, reads_overlap=reads_overlap)
 
+    except Exception as e:
+        path = os.path.join(output_dir, 'Exceptions_during_process_fastq.txt')
+        with open(path, 'w+') as f:
```
Comment (theitush) [inline processing.py]: if this works, why not just raise the exception here..?

File: processing.py [hunk 2]
```diff
@@ -234,21 +234,29 @@ def convert_fastq_to_fasta(output_dir, fastq_file):
...
+            f.write(e.__str__())
+            f.write("\n")
+        f.close()
```
Comment (theitush) [inline processing.py]: no need to close since you are using 'with'

File: runner.py [hunk 3]
```diff
@@ -85,6 +85,18 @@ def parallel_process(processing_dir, fastq_files, reference_file, quality_thresh
         raise Exception("Processing failed. This may be due to RAM overload."
                         "To avoid this try running with --max_memory max_available_RAM_in_MB")
 
+    path = os.path.join(processing_dir, 'Exceptions_during_process_fastq.txt')
+    if os.path.exists(path):
+        file_exception = open(path, 'r')
```
Comment (theitush) [inline runner.py]: please use the 'with' command

File: runner.py [hunk 4]
```diff
@@ -85,6 +85,18 @@ def parallel_process(processing_dir, fastq_files, reference_file, quality_thresh
...
+        content = file_exception.read()
+        file_exception.close()
+        exception_printed = "There were exception in the parallel processing. see " \
+                            "processing\Exceptions_during_run_blast.txt. "
```
Comment (theitush) [inline runner.py:94]: an exception could also happen in basecall so it is not specific to blast

File: runner.py [hunk 5]
```diff
@@ -275,6 +287,25 @@ def process_data(with_indels, dust, evalue, fastq_files, log, max_basecall_itera
...
 
+def validate_input(output_dir, input_dir, reference_file):
+    if os.path.exists(output_dir):
+        if os.listdir(output_dir):
+            raise Exception("Output_dir must be path to a new or empty directory!")
+    if not os.path.isfile(reference_file):
+        raise Exception("Reference_file must exist!") 
+    if not os.path.isdir(input_dir):
+        raise Exception("Input_dir must exist!")
+    input_dir_loop = os.fsencode(input_dir)
```
Comment (theitush) [inline runner.py]: why fsencode?

File: runner.py [hunk 6]
```diff
@@ -275,6 +287,25 @@ def process_data(with_indels, dust, evalue, fastq_files, log, max_basecall_itera
     return reference_file
 
 
+def validate_input(output_dir, input_dir, reference_file):
+    if os.path.exists(output_dir):
+        if os.listdir(output_dir):
+            raise Exception("Output_dir must be path to a new or empty directory!")
```
Comment (theitush) [inline runner.py]: need to merge changes I did into this

File: runner.py [hunk 7]
```diff
@@ -397,11 +428,12 @@ def create_runner_parser():
     runner(input_dir=args['input_dir'], output_dir=args['output_dir'], reference_file=args['reference_file'],
            max_basecall_iterations=int(args['max_basecall_iterations']), overlapping_reads=args['overlapping_reads'],
            quality_threshold=int(args['quality_threshold']), task=args['blast_task'], max_memory=args['max_memory'],
-           evalue=float(args['blast_evalue']), dust=args['blast_dust'], num_alignments=int(args['blast_num_alignments']),
+           evalue=float(args['blast_evalue']), dust=args['blast_dust'],
+           num_alignments=int(args['blast_num_alignments']),
```
Comment (theitush) [inline runner.py:412]: why new line?

File: runner.py [hunk 8]
```diff
@@ -397,11 +428,12 @@ def create_runner_parser():
...
            mode=args['blast_mode'], perc_identity=float(args['blast_perc_identity']), cpu_count=args['cpu_count'],
-           min_coverage=int(args['min_coverage']), db_comment=args['db_comment'], soft_masking=args['blast_soft_masking'],
+           min_coverage=int(args['min_coverage']), db_comment=args['db_comment'],
```
Comment (theitush) [inline runner.py:414]: why new line?

File: processing.py [hunk 9]
```diff
@@ -234,21 +234,29 @@ def convert_fastq_to_fasta(output_dir, fastq_file):
...
-    blast_output_file = reads_fasta_file_path + ".blast"
-    run_blast(reads_fasta=reads_fasta_file_path, reference=reference, output=blast_output_file, mode=mode, task=task,
+    try:
+        log = pipeline_logger(logger_name=f"Computation_{os.path.basename(fastq_file)}", log_folder=output_dir)
+        blast_output = os.path.join(output_dir, 'blast')
+        os.makedirs(blast_output, exist_ok=True)
+        reads_fasta_file_path = convert_fastq_to_fasta(fastq_file=fastq_file, output_dir=blast_output)
+        blast_output_file = reads_fasta_file_path + ".blast"
+        run_blast(reads_fasta=reads_fasta_file_path, reference=reference, output=blast_output_file, mode=mode, task=task,
               evalue=evalue, num_alignments=num_alignments, dust=dust, soft_masking=soft_masking, log=log,
```
Comment (theitush) [inline processing.py]: spacing

File: processing.py (hunk 1 above)
Comment (theitush) [inline processing.py]: also, better to use 'a' because 'w' will overwrite an existing file and we might have more than one error message.. no?

File: runner.py [hunk 10]
```diff
@@ -85,6 +85,18 @@ def parallel_process(processing_dir, fastq_files, reference_file, quality_thresh
         raise Exception("Processing failed. This may be due to RAM overload."
                         "To avoid this try running with --max_memory max_available_RAM_in_MB")
 
+    path = os.path.join(processing_dir, 'Exceptions_during_process_fastq.txt')
```
Comment (theitush) [inline runner.py]: filename should be lower case

File: processing.py [hunk 11]
```diff
@@ -234,21 +234,29 @@ def convert_fastq_to_fasta(output_dir, fastq_file):
 
 def process_fastq(fastq_file, reference, output_dir, quality_threshold, task, evalue, dust, num_alignments,
                   soft_masking, perc_identity, mode, reads_overlap):
-    log = pipeline_logger(logger_name=f"Computation_{os.path.basename(fastq_file)}", log_folder=output_dir)
-    blast_output = os.path.join(output_dir, 'blast')
-    os.makedirs(blast_output, exist_ok=True)
-    reads_fasta_file_path = convert_fastq_to_fasta(fastq_file=fastq_file, output_dir=blast_output)
-    blast_output_file = reads_fasta_file_path + ".blast"
-    run_blast(reads_fasta=reads_fasta_file_path, reference=reference, output=blast_output_file, mode=mode, task=task,
+    try:
```
Comment (theitush) [inline processing.py:239]: if you are using try on all of the function why not use try outside the function instead?

File: runner.py [hunk 12]
```diff
@@ -275,6 +287,25 @@ def process_data(with_indels, dust, evalue, fastq_files, log, max_basecall_itera
     return reference_file
 
 
+def validate_input(output_dir, input_dir, reference_file):
```
Comment (theitush) [inline runner.py]: this should be a separate pr as it is a different and independent feature

File: processing.py (hunk 1 above)
Comment (theitush) [inline processing.py]: aaaand why write a new file..? we have a log exactly for things like this... no?

**Review by theitush — CHANGES_REQUESTED** [review]

File: processing.py (hunk 1 above)
Comment (nitsandaniel1) [reply processing.py]: Here the exception cannot be raised- this is exactly where it gets "swallowed" instead of being raised.

Comment (theitush) [reply processing.py]: did you manage to recreate this bug outside of the pipeline?
is this a general problem with the package we are using or is it specific to something that we do with the pipeline?

Comment (nitsandaniel1) [reply processing.py]: also outside of the script using the same command the exception is not raised

File: processing.py (hunk 2 above)
Comment (nitsandaniel1) [reply processing.py]: changed

File: runner.py (hunk 3 above)
Comment (nitsandaniel1) [reply runner.py]: changed

File: runner.py (hunk 4 above)
Comment (nitsandaniel1) [reply runner.py:94]: added to be more informative- an exception will be raised also in this case.

File: runner.py (hunk 5 above)
Comment (nitsandaniel1) [reply runner.py]: to loop over the files- make sure there is a fastq file in the folder. 
removed and will be submited as another PR

File: processing.py (hunk 1 above)
Comment (nitsandaniel1) [reply processing.py]: changed

File: runner.py (hunk 10 above)
Comment (nitsandaniel1) [reply runner.py]: changed

File: processing.py (hunk 11 above)
Comment (nitsandaniel1) [reply processing.py:239]: beacuse this is specificaly where the exception can be caught in order to monitor it

File: runner.py (hunk 12 above)
Comment (nitsandaniel1) [reply runner.py]: removed and will be submited as another PR

File: processing.py (hunk 1 above)
Comment (nitsandaniel1) [reply processing.py]: this is specificly to check if there are issues during parallel computing - it there arent the file is not created.
